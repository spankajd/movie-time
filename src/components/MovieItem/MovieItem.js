import React , { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import {integerToRoman} from '../../utils/helper';
import Rating from '../Rating/Rating';

import './MovieItem.scss';

import { fetchMovieDetails, selectMovie } from '../../store/actions/movieActions';
import { selectMovieByTitle } from '../../store/selectors/movieSelectors';
import MovieDetails from '../MovieDetails/MovieDetails';

const MovieItem = React.memo(({movie, isActive=false}) => {
    const dispatch = useDispatch();
    const data = useSelector((state) => selectMovieByTitle(state,movie?.title))[0];

    useEffect( () => {
        data?.Title || dispatch(fetchMovieDetails(movie?.title));
    },[]);
    
    const onMovieSelect = (event) => {
        if(isActive) {
            dispatch(selectMovie(null));
        } else {
            dispatch(selectMovie(data));
        }
    }

    return <> 
    <div className={`movie-item ${isActive && 'active'}`} onClick={onMovieSelect} data-testid="movie-item">
        <div className="movie-id">Episode {movie?.episode_id}</div>
        <div className="movie-title">Episode {integerToRoman(movie?.episode_id)} - {movie?.title}</div>
        <div className="movie-rating"><Rating rating={parseInt(movie?.imdbRating)} /></div>
        <div className="movie-release-date">{movie?.release_date}</div>
    </div>
    {isActive && <MovieDetails />}
    </>
});

export default MovieItem;