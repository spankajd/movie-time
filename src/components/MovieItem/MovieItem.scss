.movie-item {
    display: flex;
    width: calc(100% - 20px);
    padding: 10px;
    cursor: pointer;
    background-color:#e9f5ff;
    
    &:hover {
        background-color:#d4e8fb;
    }
    &.active {
        background-color:#d4e8fb;
    }

    .movie-id {
        width: 15%;
    }
    .movie-title {
        width: 45%;
    }
    .movie-rating {
        width: 25%;
    }
    .movie-release-date {
        width: 15%;
        text-align: right;
    }

  @media (max-width: 768px) {
    .movie-id {
        display: none;
    }
    .movie-rating {
        display: none;
    }
    .movie-title {
        width: 75%;
    }
    .movie-release-date {
        width: 25%;
    }
  }
  @media (min-width: 768px) and (max-width: 1024px) {
    .movie-id,
    .movie-rating {
        display: none;
    }
    .movie-title {
        width: 60%;
    }
    .movie-release-date {
        width: 40%;
    }
  }
}
