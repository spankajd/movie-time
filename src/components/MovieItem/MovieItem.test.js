import { fireEvent, screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import { renderWithProviders } from "../../utils/test-utlis";
import MovieItem from "./MovieItem";

describe("MovieItem Component", () => {

    it("matches snapshot", () => {
        const { asFragment } = renderWithProviders(<MovieItem movie={{title:'Test'}} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders MovieItem", () => {
        renderWithProviders(
            <MovieItem movie={{title:'Test'}}/>
        );
        const element = screen.getByText(/Test/i);
        expect(element).toBeInTheDocument();
    });


    it("renders MovieItem with isActive as true", () => {
        renderWithProviders(
            <MovieItem movie={{title:'Test'}} isActive={true}/>
        );
        const element = screen.getByText(/Test/i);
        expect(element).toBeInTheDocument();
    });

    it("renders MovieItem and trigger click event", () => {
        renderWithProviders(
            <MovieItem movie={{title:'Test'}} isActive={true}/>
        );
        fireEvent.click(screen.getByTestId('movie-item'));
        
        const element = screen.getByText(/Please select movie from the list/i);
        expect(element).toBeInTheDocument();
    });

    it("renders MovieItem and trigger click event two times", () => {
        renderWithProviders(
            <MovieItem movie={{title:'Test'}}/>
        );
        fireEvent.click(screen.getByTestId('movie-item'));
        fireEvent.click(screen.getByTestId('movie-item'));

        const element = screen.getByText(/Test/i);
        expect(element).toBeInTheDocument();
    });
});