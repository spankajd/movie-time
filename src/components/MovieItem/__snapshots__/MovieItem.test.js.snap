// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-item false"
    data-testid="movie-item"
  >
    <div
      class="movie-id"
    >
      Episode 
    </div>
    <div
      class="movie-title"
    >
      Episode  - Test
    </div>
    <div
      class="movie-rating"
    >
      <div
        class="rating"
      >
        N/A
      </div>
    </div>
    <div
      class="movie-release-date"
    />
  </div>
</DocumentFragment>
`;
